ui/prometheus-plugin/src/plugins/PrometheusTimeSeriesQueryEditor.tsx

```diff
@@ -0,0 +1,35 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { ChangeEvent } from 'react';
+import produce from 'immer';
+import { TextField } from '@mui/material';
+import { OptionsEditorProps } from '@perses-dev/plugin-system';
+import { PrometheusTimeSeriesQuerySpec } from './time-series-query';
+
+export type PrometheusTimeSeriesQueryEditorProps = OptionsEditorProps<PrometheusTimeSeriesQuerySpec>;
+
+export function PrometheusTimeSeriesQueryEditor(props: PrometheusTimeSeriesQueryEditorProps) {
+  const { onChange, value } = props;
+  const { query } = value;
+
+  const handleQueryChange = (e: ChangeEvent<HTMLInputElement>) => {
+    onChange(
+      produce(value, (draft: PrometheusTimeSeriesQuerySpec) => {
```

Comment:
Is immer tied to the plugin system? or can developers decide how they want to update the state in their plugin and the plugin just think it's just some normal JS value? I think this is cool

Reply:
I left a TODO to figure out whether we should expose it as part of the plugin system:

https://github.com/perses/perses/blob/47bc3d1a7d3bfa06e38fd7ed6f2e1812bc89a4ab/ui/dashboards/src/components/PanelDrawer/panel-editor-model.ts#L31-L37

But this is a totally cool way to do it without exposing it. ❤️

Reply:
This is making me realize that I used `immer` without installing `immer` in either package (`panels-plugin` or `prometheus-plugin`)... but it worked?

I'm guessing it's because they both depend on `plugin-system` and that installs `immer`.

But is that the right way to do things? Normally, if I install some library, I don't expect to be able to use whatever libraries it has installed too...

If I'm doing this, should I install `immer` directly in `panels-plugin` and `prometheus-plugin`?

Reply:
Yeah good catch, I’d add it in those package.json’s too to be safe

Reply:
Just added to `panels-plugin` and `prometheus-plugin`! (in this PR)